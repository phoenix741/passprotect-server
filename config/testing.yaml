config:
    mongodb:
        host: 'mongodb://localhost:27017/passprotect'
        user: ~
        password: ~
        options: ~
