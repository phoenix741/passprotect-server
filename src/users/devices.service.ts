import { Model } from 'mongoose';
import { Injectable, UnauthorizedException } from '@nestjs/common';
import { UserEntity } from './models/user.entity';
import { DeviceEntity } from './models/user.entity';
import { UsersService } from './users.service';

@Injectable()
export class DevicesService {
  constructor(private readonly userService: UsersService) {}

  async findAll(userId: string): Promise<DeviceEntity[]> {
    if (userId) {
      const user = await this.userService.findById(userId);
      return user.devices;
    }
    return [];
  }

  async verifyDevice(user: UserEntity, deviceId: string): Promise<UserEntity> {
    const isValid = user.devices.find(d => d.id === deviceId);
    if (!isValid) {
      throw new UnauthorizedException('error.user.404.userNotFound');
    }
    return user;
  }

  async registerDevice(userId: string, device: DeviceEntity): Promise<DeviceEntity> {
    const user = await this.userService.findById(userId);
    if (!user) {
      throw new UnauthorizedException('error.user.404.userNotFound');
    }

    const found = user.devices.find(d => d.id === device.id);
    if (!found) {
      user.devices.push(device);
      await user.save();
    }
    return found || device;
  }

  async removeDevice(userId: string, deviceId: string): Promise<void> {
    const user = await this.userService.findById(userId);
    if (!user) {
      throw new UnauthorizedException('error.user.404.userNotFound');
    }

    user.devices = user.devices.filter(d => d.id !== deviceId);
    await user.save();
  }
}
