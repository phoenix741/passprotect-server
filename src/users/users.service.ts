import { Model } from 'mongoose';
import { Injectable, UnauthorizedException } from '@nestjs/common';
import { InjectModel } from '@nestjs/mongoose';
import { UserToRegister } from './models/user.models';
import { UserEntity } from './models/user.entity';
import { CryptoService } from '../shared/services/crypto.service';
import { DeviceEntity } from '../users/models/user.entity';

@Injectable()
export class UsersService {
  constructor(private readonly cryptoService: CryptoService, @InjectModel('User') private readonly userModel: Model<UserEntity>) {}

  async findAll(): Promise<UserEntity[]> {
    return await this.userModel.find().exec();
  }

  async findById(id: string): Promise<UserEntity | null> {
    return id && (await this.userModel.findById(id.toLowerCase()));
  }

  async registerUser(user: UserToRegister): Promise<UserEntity> {
    user.password = await this.cryptoService.hashPassword(user.password);

    return await this.userModel.create(user);
  }

  async verifyPassword(user: UserEntity, password: string): Promise<UserEntity> {
    const isValid = await this.cryptoService.checkPassword(password, user.password);
    if (!isValid) {
      throw new UnauthorizedException('error.user.404.userNotFound');
    }
    return user;
  }
}
