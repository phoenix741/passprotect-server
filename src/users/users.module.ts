import { Module } from '@nestjs/common';
import { MongooseModule } from '@nestjs/mongoose';
import { UserSchema } from './schemas/user.schema';
import { UsersService } from './users.service';
import { SharedModule } from '../shared/shared.module';
import { DevicesService } from './devices.service';

@Module({
  imports: [MongooseModule.forFeature([{ name: 'User', schema: UserSchema }]), SharedModule],
  providers: [UsersService, DevicesService],
  exports: [UsersService, DevicesService],
})
export class UsersModule {}
