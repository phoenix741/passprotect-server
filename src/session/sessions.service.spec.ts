import { Test } from '@nestjs/testing';
import { JwtService } from '@nestjs/jwt';
import { UsersService } from '../users/users.service';
import { SessionsService } from './sessions.service';

describe('SessionsService', () => {
  let userService;
  let jwtService;
  let sessionsService;

  const userEntity = {
    _id: 'username',
    password: 'password',
    encryption: {
      salt: Buffer.from('salt'),
      iv: Buffer.from('iv'),
      content: Buffer.from('content'),
      authTag: Buffer.from('authTag'),
    },
  };

  const jwtToken = 'myJwtToken';

  beforeEach(async () => {
    userService = {
      findById: jest.fn().mockImplementation(() => userEntity),
      verifyPassword: jest.fn().mockImplementation(() => userEntity),
    };
    jwtService = {
      sign: jest.fn().mockImplementation(() => jwtToken),
    };

    const module = await Test.createTestingModule({
      providers: [
        SessionsService,
        {
          provide: UsersService,
          useValue: userService,
        },
        {
          provide: JwtService,
          useValue: jwtService,
        },
      ],
    }).compile();

    sessionsService = module.get<SessionsService>(SessionsService);
  });

  describe('signIn', () => {
    it('user not found', async () => {
      userService.findById.mockImplementation(() => null);
      await expect(sessionsService.signIn('username', 'password', 'fingerprint')).rejects.toThrowErrorMatchingSnapshot();
    });

    it('wrong password', async () => {
      userService.verifyPassword.mockImplementation(async () => {
        throw new Error('wrong password');
      });
      await expect(sessionsService.signIn('username', 'password', 'fingerprint')).rejects.toThrowErrorMatchingSnapshot();
    });

    it('user found', async () => {
      expect(await sessionsService.signIn('username', 'password', 'fingerprint')).toMatchSnapshot('user sign in');
      expect(jwtService.sign).toHaveBeenCalledWith({ sub: 'username', issuer: 'fingerprint' }, {expiresIn: '15m', jwtid: expect.any(String)});
    });
  });

  describe('validateUser', () => {
    it('success', async () => {
      await expect(await sessionsService.validateUser({ sub: 'username' })).toMatchSnapshot('user validation');
      expect(userService.findById).toHaveBeenCalledWith('username');
    });
  });
});
