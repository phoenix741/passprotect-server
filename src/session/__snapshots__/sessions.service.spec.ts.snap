// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SessionsService signIn user found: user sign in 1`] = `
Object {
  "jwtToken": "myJwtToken",
  "user": Object {
    "_id": "username",
    "encryption": Object {
      "authTag": Object {
        "data": Array [
          97,
          117,
          116,
          104,
          84,
          97,
          103,
        ],
        "type": "Buffer",
      },
      "content": Object {
        "data": Array [
          99,
          111,
          110,
          116,
          101,
          110,
          116,
        ],
        "type": "Buffer",
      },
      "iv": Object {
        "data": Array [
          105,
          118,
        ],
        "type": "Buffer",
      },
      "salt": Object {
        "data": Array [
          115,
          97,
          108,
          116,
        ],
        "type": "Buffer",
      },
    },
    "password": "password",
  },
}
`;

exports[`SessionsService signIn user not found 1`] = `
Object {
  "error": "Unauthorized",
  "message": "error.user.404.userNotFound",
  "statusCode": 401,
}
`;

exports[`SessionsService signIn wrong password 1`] = `"wrong password"`;

exports[`SessionsService validateUser success: user validation 1`] = `Promise {}`;
